label_candidates/label_candidates.py

```diff
@@ -0,0 +1,254 @@
+"""
+
+candidate = pd.read_sql(sql, database_str)
+print(candidate.shape)
+candidate.head()
+# -
+
+candidate = candidate.assign(
+    word=lambda x: x.apply(lambda row: row.word.replace("'", "").split("|"), axis=1),
+    pos_tag=lambda x: x.apply(
+        lambda row: row.pos_tag.replace("'", "").split("|"), axis=1
+    ),
+)
+print(candidate.shape)
+candidate.head()
+
+# +
+lf_columns = list(DaG["distant_supervision"].values()) + list(
+    DaG["text_patterns"].values()
+)
+
+lf_column_names = [col.name for col in lf_columns]
+# -
+
+lf_applier = PandasLFApplier(lfs=lf_columns)
+labels = lf_applier.apply(df=candidate)
+
+train_lfs_df = pd.DataFrame(labels, columns=lf_column_names).assign(
+    candidate_id=candidate.candidate_id.tolist()
+)
+print(train_lfs_df.shape)
+train_lfs_df.head()
+
+(train_lfs_df.to_csv("output/dg_full_text_train_candidates.tsv", sep="\t", index=False))
+
+# ## Dev Set
+
+annotated_df = pd.read_csv(
+    "../annotation_conversion/output/mapped_disease_gene_candidates.tsv", sep="\t"
+)
+annotated_df.head()
+
+
+def char_to_word(row):
+    char_word_mapper = [
+        (idx, int(tok)) for idx, tok in enumerate(row["char_offsets"].split("|"))
+    ]
+    gene_end = -1
+    gene_start = -1
+    disease_end = -1
+    disease_start = -1
+    for word in char_word_mapper:
+        if word[1] >= row["disease_start"] and word[1] <= row["disease_end"]:
+            if word[1] == row["disease_start"]:
+                disease_start = word[0]
+            else:
+                disease_end = word[0]
+        if word[1] >= row["gene_start"] and word[1] <= row["gene_end"]:
+            if word[1] == row["gene_start"]:
+                gene_start = word[0]
+            else:
+                gene_end = word[0]
+
+    return (
+        disease_start,
+        disease_end if disease_end != -1 else disease_start + 1,
+        gene_start,
+        gene_end if gene_end != -1 else gene_start + 1,
+    )
+
+
+fixed_map_df = pd.DataFrame.from_dict(
+    dict(
+        zip(
+            ["disease_start", "disease_end", "gene_start", "gene_end"],
+            zip(*annotated_df.apply(char_to_word, axis=1).tolist()),
+        )
+    )
+)
+fixed_map_df
+
+fixed_annotated_df = (
+    annotated_df.assign(
+        disease_start=fixed_map_df.disease_start.tolist(),
+        disease_end=fixed_map_df.disease_end.tolist(),
+        gene_start=fixed_map_df.gene_start.tolist(),
+        gene_end=fixed_map_df.gene_end.tolist(),
+    )
+    .rename(
+        index=str,
+        columns={
+            "Disease_cid": "disease_cid",
+            "Gene_cid": "gene_cid",
+            "words": "word",
+            "lemmas": "lemma",
+            "pos_tags": "pos_tag",
+        },
+    )
+    .assign(
+        lemma=lambda x: x.lemma.apply(lambda s: s.split("|")),
+        word=lambda x: x.word.apply(lambda s: s.split("|")),
+        pos_tag=lambda x: x.pos_tag.apply(lambda s: s.split("|")),
+    )
+)
+fixed_annotated_df
```

Comment:
Is it okay to be getting NaN sentence IDs here?

Reply:
Yeah its fine. As those NaNs are the sentences that failed to be processed via my pipeline. Processing full text and abstracts is not a clean process.